### PR TITLE
server: start to listen after init stats complete

### LINKED ISSUE #51473: listen the port before the init stats
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

add the ```time.Sleep``` into the checkpoint of ```force_init_stats```.

and then ```nc 127.0.0.1 4000```

### 2. What did you expect to see? (Required)

cannot connect

### 3. What did you see instead (Required)

can connect

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #51473

Problem Summary:

### What changed and how does it work?

Now the checkpoint of ```force-init-stats``` is before the TCP ```accept``` handler. it should be before establishing the TCP listener.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

Add the sleep into the checkpoint of ```force-init-stats```. then we cannot connect it.

we cannot add it to UT. because it takes too much time for UT to sleep duration running.

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: cmd/tidb-server/main.go [hunk 1]
```diff
@@ -302,11 +302,6 @@ func main() {
 	storage, dom := createStoreAndDomain(keyspaceName)
 	svr := createServer(storage, dom)
 
-	// Register error API is not thread-safe, the caller MUST NOT register errors after initialization.
-	// To prevent misuse, set a flag to indicate that register new error will panic immediately.
-	// For regression of issue like https://github.com/pingcap/tidb/issues/28190
-	terror.RegisterFinish()
```
Comment (hawkingrei) [inline cmd/tidb-server/main.go:308]: It is moved into ```server.Run()```.

File: pkg/util/channel/channel.go [hunk 2]
```diff
@@ -20,3 +20,14 @@ func Clear[T any, V chan T | <-chan T](ch V) {
 	for range ch {
 	}
 }
+
+// IsClosed is to check if the channel is closed
+func IsClosed[T any, V chan T | <-chan T](ch V) bool {
```
Comment (tiancaiamao) [inline pkg/util/channel/channel.go:25]: Why not just `V chan T` here?

Comment (hawkingrei) [reply pkg/util/channel/channel.go:25]: It will report that T is undefined.

File: pkg/util/channel/channel.go [hunk 3]
```diff
@@ -20,3 +20,14 @@ func Clear[T any, V chan T | <-chan T](ch V) {
...
+	select {
+	case <-ch:
```
Comment (tiancaiamao) [inline pkg/util/channel/channel.go:27]: This is more general and correct:

```
case _, ok := <-ch:
    return !ok
```

Over all, I dislike introducing additional functions where we can just use language builtin support.

Comment (hawkingrei) [reply pkg/util/channel/channel.go:27]: Sure, I find it is a bad function. it will remove it.

Comment (tiancaiamao) [reply pkg/util/channel/channel.go:27]: Try to test it:

```
ch := make(chan int)
ch <- 1
util.IsClosed(ch)
```

**Review by tiancaiamao — APPROVED** [review]

**Review by elsa0520 — APPROVED** [review]
LGTM

**Review by 3pointer — APPROVED** [review]
BR part LGTM

**Pull request merged**